src/document.rs

```diff
@@ -123,16 +121,16 @@ pub(crate) fn extract_value_for_type(
             }
 
             Value::Object(
-                any.downcast::<PyDict>()
-                    .map(|dict| pythonize::depythonize(dict))
-                    .map_err(to_pyerr_for_type("Json", field_name, any))?
+                any.clone().downcast_into::<PyDict>()
+                    .map(|dict| pythonize::depythonize_bound(dict.clone().into_any()))
+                    .map_err(to_pyerr_for_type("Json", field_name, any.clone()))?
```

Comment:
(Why are there two calls to `map_err` here? Look at the LHS, they're both the same...?)

Reply:
I think one is handling the error from `downcast` and one is handling the error from `depythonize`.

I think the code could be made clearer using either `and_then`

```rust
 any.clone()
    .downcast_into::<PyDict>()
    .map_err(to_pyerr_for_type("Json", field_name, any.clone())
    .and_then(|dict| {
       pythonize::depythonize_bound(dict.into_any())
       .map_err(to_pyerr_for_type("Json", field_name, any.clone())
    })?;
```

or a temporary binding

```rust
let dict = any.clone().downcast_into::<PyDict>().map_err(to_pyerr_for_type("Json", field_name, any.clone()))?;
let obj =  pythonize::depythonize_bound(dict.into_any()).map_err(to_pyerr_for_type("Json", field_name, any))?

Value::Object(obj)
```

but both variants still need two calls to `map_err` as the source `Result` type is different.

(I think the `clone` on the `dict` should be unnecessary as it is not used elsewhere but I have not actually tried it with the borrow checker.)

Reply:
I mentioned in the other comment, seems to me `depythonize_bound` doesn't take a reference, it wants the owned value.

Reply:
> I think the code could be made clearer

Your `and_then` example looks much better thank you 😄